uat/testing-features/src/main/java/com/aws/greengrass/steps/AssertionSteps.java

```diff
@@ -0,0 +1,228 @@
+    public void assertionsWithContextTimeout(int assertionCount, String context, String message,
+                                                 long timeoutSeconds) throws Throwable {
+
+        assertionsWithContextTimeout(assertionCount, scenarioContext.applyInline(context),
+                message, timeoutSeconds, true);
+    }
+
+    @SuppressWarnings("PMD.AvoidCatchingThrowable")
+    private void assertionsWithContextTimeout(int assertionCount, String context, String message,
+                                              long timeoutSeconds, boolean failOnTooManyAssertions)
+            throws Throwable {
+        String exceptionMessage = (message == null)
+                ? String.format("did not receive %d assertions for '%s' in time (%ds)", assertionCount,
+                        context, timeoutSeconds)
+                : String.format("did not receive %d assertions for '%s' with message '%s' in time (%ds)",
+                assertionCount, context, message, timeoutSeconds);
+
+        class Wrapper<T> {
+            @Getter
+            @Setter
+            private T wrapped;
+        }
+
+        Wrapper<Throwable> wrapper = new Wrapper<>();
+        boolean finished = waits.untilTrue(() -> {
+            try {
+                return assertionsReceived(assertionCount, context, message, failOnTooManyAssertions, exceptionMessage);
+            } catch (Throwable e) {
+                wrapper.setWrapped(e);
+                return true;
+            }
+        }, (int) timeoutSeconds, TimeUnit.SECONDS);
+
+        if (!finished) {
+            throw new TimeoutException("Timeout: " + exceptionMessage);
+        }
+
+        Throwable e = wrapper.getWrapped();
+        if (e != null) {
+            throw e;
+        }
+    }
+
+    private boolean assertionsReceived(int assertionCount, String context, String message,
+                                          boolean failOnTooManyAssertions, String exceptionMessage) {
+        List<Assertion> assertions = assertionList.stream()
+                .filter(assertion -> assertion.getContext() != null)
+                .filter(assertion -> context.equalsIgnoreCase(assertion.getContext()))
+                .filter(assertion -> message == null || message.equalsIgnoreCase(assertion.getMessage()))
+                .collect(Collectors.toList());
+        for (Assertion assertion : assertions) {
+            if (!assertion.isSuccess()) {
+                fail(String.format("Assertion with message %s and context %s FAILED", assertion.getMessage(),
+                        assertion.getContext()));
+            }
+        }
+        if (failOnTooManyAssertions && assertions.size() > assertionCount) {
+            fail(exceptionMessage);
+        }
+        return assertions.size() >= assertionCount;
+    }
+
+    @After
+    @Override
+    public void close() {
+        if (server != null) {
+            server.stop(0);
+            ((ExecutorService) server.getExecutor()).shutdownNow();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Call `shutdown()` to reject any incoming tasks and `awaitTermination()` to wait for existing tasks to complete or until a timeout is reached. To stop the execution and processing of tasks, if necessary, call `shutdownNow()`. Check the return value of `awaitTermination()` and handle when the method times out while waiting for other threads to stop executing. See the following diffs on GitHub that make this code change: [diff1](https://github.com/glowroot/glowroot/commit/81de5a3699c59e0cf98d485c380c1034e6cd28a5#diff-c608d2c40790b0e0e5228e6647f69729caa756bbe9127169139406314b80897eL143), [diff2](https://github.com/apache/directory-kerby/commit/3ed0e7cd3b8a0c3ce29293153e4e72f422f0637f#diff-eb42f072849f520563ec2a686789a836195f18075cc222ef355c6ea710b0a360L84)

Learn more: [ExecutorService in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html)